prow/cmd/webhook-server/helpers_test.go

```diff
@@ -133,15 +144,20 @@ func TestUpdateSecret(t *testing.T) {
 			}
 
 			gotCert, gotPrivKey, gotCaPem, gotErr := updateSecret(f, ctx, clientoptions)
-			if tc.name == "noDnsNames" && gotErr == nil {
-				t.Fatalf("unexpected error %v", gotErr)
-			} else if tc.name != "noDnsNames" && gotErr != nil {
-				t.Fatalf("unexpected error %v", gotErr)
+			if tc.wantErr {
+				if gotErr == nil {
```

Comment:
Instead of `tc.wantErr` as a bool, why not (1) have `tc.expectedErr` be a nil or non-nil type, and just do a single `if tc.wantErr != gotErr { t.Fatalf(...) }`?

Reply:
strict equality comparison for errors in Go is not very efficient, especially when there are nested inner errors. So if we do want to check which type of error is returned, we could either do inner error check such as `errors.Is()`, but it requires all inner errors are wrapped with `%w`, which is not the case everywhere. Or we could do partial string matching, but the logic will also be multi lines, something like:
```
if gotErr != nil && wantErr != nil && strings.Contains(gotErr.Error, wantErr.Error) {
  t.Fatal()
}

Reply:
Oops, I meant to say `if tc.exectedErr != gotErr { ... }`.

> not very efficient

I don't think efficiency is a concern for unit tests.

> Or we could do partial string matching, but the logic will also be multi lines

How about something like https://pkg.go.dev/github.com/stretchr/testify@v1.8.0/assert#EqualError ? Looks like we already use this library in other parts of the codebase.

At the very least I think we should change line 153 to additionally log a statement saying "aborting test" before returning.

Reply:
I agree that EqualError seems very useful, but there is not too much value to check here. added a log at line 153.